packages/wp-now/src/wp-now.ts

```diff
@@ -0,0 +1,299 @@
+				isStaticFilePath: (path) => {
+					try {
+						const fullPath = this.options.documentRoot + path;
+						return (
+							this.php.fileExists(fullPath) &&
+							!this.php.isDir(fullPath) &&
+							!seemsLikeAPHPFile(fullPath)
+						);
+					} catch (e) {
+						console.error(e);
+						return false;
+					}
+				},
+			},
+		});
+		this.php.mkdirTree(documentRoot);
+		this.php.chdir(documentRoot);
+		this.php.writeFile(
+			`${documentRoot}/index.php`,
+			`<?php echo 'Hello wp-now!';`
+		);
+	}
+
+	mountWordpress(fileName = 'latest') {
+		const { documentRoot } = this.options;
+		const root = path.join(WORDPRESS_ZIPS_PATH, fileName, 'wordpress');
+		this.php.mount(
+			{
+				root,
+			},
+			documentRoot
+		);
+		this.php.writeFile(
+			`${documentRoot}/wp-config.php`,
+			this.php.readFileAsText(`${documentRoot}/wp-config-sample.php`)
+		);
+		this.updateFile(
+			`${documentRoot}/wp-config.php`,
+			(contents) =>
+				`<?php
+          define('WP_HOME', "${this.options.absoluteUrl}");
+          define('WP_SITEURL', "${this.options.absoluteUrl}");
+          ?>${contents}`
+		);
+		this.php.mkdirTree(`${documentRoot}/wp-content/mu-plugins`);
+		this.php.writeFile(
+			`${documentRoot}/wp-content/mu-plugins/0-allow-wp-org.php`,
+			`<?php
+      // Needed because gethostbyname( 'wordpress.org' ) returns
+      // a private network IP address for some reason.
+      add_filter( 'allowed_redirect_hosts', function( $deprecated = '' ) {
+          return array(
+              'wordpress.org',
+              'api.wordpress.org',
+              'downloads.wordpress.org',
+          );
+      } );`
+		);
+	}
+
+	async runCode(code) {
+		const result = await this.php.run({
+			code,
+		});
+		console.log(result.text);
+		return result;
+	}
+
+	mountSqlite() {
+		const { documentRoot } = this.options;
+		const sqlitePluginPath = `${documentRoot}/wp-content/plugins/${SQLITE_FILENAME}`;
+		this.php.mkdirTree(sqlitePluginPath);
+		this.php.mount(SQLITE_PATH, sqlitePluginPath);
+		this.php.writeFile(
+			`${documentRoot}/wp-content/db.php`,
+			this.php
+				.readFileAsText(`${sqlitePluginPath}/db.copy`)
+				.replace(
+					/\{SQLITE_IMPLEMENTATION_FOLDER_PATH\}/g,
+					`${documentRoot}/wp-content/plugins/${SQLITE_FILENAME}`
+				)
+				.replace(/\{SQLITE_PLUGIN\}/g, SQLITE_FILENAME)
+		);
+	}
+
+	static #getWpContentHomePath(projectPath: string) {
+		const basename = path.basename(projectPath);
+		const directoryHash = crypto
+			.createHash('sha1')
+			.update(projectPath)
+			.digest('hex');
+		return path.join(
+			WP_NOW_PATH,
+			'wp-content',
+			`${basename}-${directoryHash}`
+		);
+	}
+
+	static #isPluginDirectory(projectPath: string): Boolean {
+		const files = fs.readdirSync(projectPath);
+		for (const file of files) {
+			if (file.endsWith('.php')) {
+				const fileContent = fs.readFileSync(
+					path.join(projectPath, file),
+					'utf8'
+				);
+				if (fileContent.includes('Plugin Name:')) {
+					return true;
+				}
+			}
```

Comment:
in some cases this could end up reading _considerable_ amounts of data. we might want to watch it at a minimum and consider pruning what we read. for example, we need not read in the entire file into memory when scanning for the PHP comment. we might also be okay to draw a line at so many bytes into a file and say "if there's no plugin name before now then we skip it"

both `fs.readFile` and `fs.read` give us the ability to load chunks into memory while searching and abort when we've decided it's not a plugin. with `fs.readFile` we do that within the callback and might need to use the `AbortController`, but with `fs.read` we can specify directly how much we're willing to examine.

e.g. if there's no `Plugin Name:` in the first 4KB of a file we scrap it. altneratively we might do better by doing primitive parsing of the PHP and look for an opening `^/**` up to the end of the comment `*/`, and search only within there for `\s*\*\s*Plugin Name:\s+`

Reply:
Great suggestion, I've captured this optimization in this issue https://github.com/WordPress/wordpress-playground/issues/227